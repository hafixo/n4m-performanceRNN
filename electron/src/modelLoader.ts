import manifest from './weights_manifest.json'
import * as tf from '@tensorflow/tfjs-node';
import { WeightsManifestConfig } from '@tensorflow/tfjs-core/dist/io/io';
const CHECKPOINT_URL = 'https://storage.googleapis.com/' +
    'download.magenta.tensorflow.org/models/performance_rnn/tfjs'

export async function loadModel() {
  const weights = await tf.io.loadWeights(manifest as WeightsManifestConfig, CHECKPOINT_URL);
  const models = modelFactory(weights);
  return models
}

export type Models = ReturnType<typeof modelFactory>

function modelFactory(vars: {[varName: string]: tf.Tensor}) {
  const lstmKernel1 =
        vars['rnn/multi_rnn_cell/cell_0/basic_lstm_cell/kernel'] as
        tf.Tensor2D;
  const lstmBias1 = vars['rnn/multi_rnn_cell/cell_0/basic_lstm_cell/bias'] as
        tf.Tensor1D;
  const lstmKernel2 =
        vars['rnn/multi_rnn_cell/cell_1/basic_lstm_cell/kernel'] as
        tf.Tensor2D;
  const lstmBias2 = vars['rnn/multi_rnn_cell/cell_1/basic_lstm_cell/bias'] as
        tf.Tensor1D;
  const lstmKernel3 =
        vars['rnn/multi_rnn_cell/cell_2/basic_lstm_cell/kernel'] as
        tf.Tensor2D;
  const lstmBias3 = vars['rnn/multi_rnn_cell/cell_2/basic_lstm_cell/bias'] as
        tf.Tensor1D;
  const fcB = vars['fully_connected/biases'] as tf.Tensor1D;
  const fcW = vars['fully_connected/weights'] as tf.Tensor2D;

  return {
    lstmKernel1,
    lstmBias1,
    lstmKernel2,
    lstmBias2,
    lstmKernel3,
    lstmBias3,
    fcB,
    fcW
  }
}