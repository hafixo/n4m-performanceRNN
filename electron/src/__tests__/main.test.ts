import App, { getInitialWeights } from '../main'
import { loadModel, Models } from '../modelLoader'

describe('App', () => {
  let app: App;
  let models: Models;

  beforeAll(async () => {
    models = await loadModel()
  })

  beforeEach(() => {
    app = new App(models);
  })

  describe('getInitialWeights', () => {
    describe('with valid models', () => {
      it('returns valid shape', () => {
        const w = getInitialWeights(models)
        expect(w).toMatchSnapshot();
      })
    })
    describe('with invalid models', () => {
      it('throws', () => {
        const invalidModels = {} as Models
        expect(() => {
          getInitialWeights(invalidModels)
        }).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('constructor', () => {
    it('app is defined', () => {
      expect(App).toBeDefined()
      expect(app).toBeInstanceOf(App)
    })
    it('models are set', () => {
      expect(app.models).toBeDefined()
      expect(app.models).toMatchSnapshot()
    })
    it('conditioned is set false', () => {
      expect(app.conditioned).toBe(false)
    })
  })

  describe('initialize', () => {
    it('sets loop id at 0', async () => {
      app.currentLoopId = 100;
      expect(app.currentLoopId).toBe(100)
      await app.initialize()
      expect(app.currentLoopId).toBe(0)
    })
  })
})
