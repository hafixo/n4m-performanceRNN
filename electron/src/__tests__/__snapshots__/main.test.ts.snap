// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App constructor models are set 1`] = `
Object {
  "fcB": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 6,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "1",
    "shape": Array [
      388,
    ],
    "size": 388,
    "strides": Array [],
  },
  "fcW": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 3,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      512,
      388,
    ],
    "size": 198656,
    "strides": Array [
      388,
    ],
  },
  "lstmBias1": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 7,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "1",
    "shape": Array [
      2048,
    ],
    "size": 2048,
    "strides": Array [],
  },
  "lstmBias2": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 2,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "1",
    "shape": Array [
      2048,
    ],
    "size": 2048,
    "strides": Array [],
  },
  "lstmBias3": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 5,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "1",
    "shape": Array [
      2048,
    ],
    "size": 2048,
    "strides": Array [],
  },
  "lstmKernel1": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 1,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      921,
      2048,
    ],
    "size": 1886208,
    "strides": Array [
      2048,
    ],
  },
  "lstmKernel2": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 4,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      1024,
      2048,
    ],
    "size": 2097152,
    "strides": Array [
      2048,
    ],
  },
  "lstmKernel3": Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 8,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      1024,
      2048,
    ],
    "size": 2097152,
    "strides": Array [
      2048,
    ],
  },
}
`;

exports[`App getInitialWeights with invalid models throws 1`] = `"Invalid Model shape. Model must have the following properties. lstmBias1, lstmBias2, lstmBias3"`;

exports[`App getInitialWeights with valid models returns valid shape 1`] = `
Array [
  Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 9,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      1,
      512,
    ],
    "size": 512,
    "strides": Array [
      512,
    ],
  },
  Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 10,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      1,
      512,
    ],
    "size": 512,
    "strides": Array [
      512,
    ],
  },
  Tensor {
    "dataId": Object {},
    "dtype": "float32",
    "id": 11,
    "isDisposedInternal": false,
    "kept": false,
    "rankType": "2",
    "shape": Array [
      1,
      512,
    ],
    "size": 512,
    "strides": Array [
      512,
    ],
  },
]
`;
